File: app/src/screens/SecuritySettingsScreen.js
// GÜVENLİK AYARLARI — şifre değiştir · e-posta değiştir · diğer oturumları
// kapat · son güvenlik olayları.
//
//   • Şifre ve e-posta değişikliği MEVCUT ŞİFRE ile yeniden doğrulama ister
//     (sunucu da ayrıca zorunlu kılar — arayüz atlatılamaz).
//   • Şifre değişince diğer cihazlardaki oturumlar sunucuda kapatılır.
//   • Olay listesinde şifre/token benzeri hassas veri YOKTUR (sunucu süzer).
import React, { useEffect, useState } from 'react';
import { View, Text, TouchableOpacity, StyleSheet, ScrollView, ActivityIndicator, Switch, TextInput } from 'react-native';
import { colors, spacing, radius } from '../theme';
import ScreenBackdrop from '../components/ScreenBackdrop';
import { api } from '../api';
import { useAuth, refreshUser } from '../auth';
import { PasswordField } from './AuthScreens';
import {
  biometricsSupported, isBioLockEnabled, setBioLockEnabled, authenticate,
} from '../security/biometricLock';

function Card({ title, children }) {
  return (
    <View style={styles.card}>
      <Text style={styles.cardTitle}>{title}</Text>
      {children}
    </View>
  );
}

function Feedback({ ok, err }) {
  if (ok) return <Text style={styles.ok}>{ok}</Text>;
  if (err) return <Text style={styles.err}>{err}</Text>;
  return null;
}

// Biyometrik kilit — yalnız destekleyen cihazlarda görünür (web'de hiç).
// Aç/kapat için önce cihazın biyometrik doğrulaması istenir; böylece kilidi
// yalnız cihaz sahibi değiştirebilir. Biyometrik veri uygulamaya girmez.
function BiometricCard() {
  const [supported, setSupported] = useState(false);
  const [enabled, setEnabled] = useState(false);
  const [busy, setBusy] = useState(false);
  const [msg, setMsg] = useState(null);
  useEffect(() => {
    let alive = true;
    (async () => {
      const s = await biometricsSupported();
      if (!alive) return;
      setSupported(s);
      if (s) setEnabled(await isBioLockEnabled());
    })();
    return () => { alive = false; };
  }, []);

  if (!supported) return null;

  const toggle = async (next) => {
    setBusy(true); setMsg(null);
    const outcome = await authenticate(); // değişiklik için kimlik doğrulaması şart
    if (outcome !== 'unlocked') {
      setMsg('Doğrulama başarısız — ayar değiştirilmedi.');
      setBusy(false);
      return;
    }
    await setBioLockEnabled(next);
    setEnabled(next);
    setMsg(next
      ? 'Biyometrik kilit açıldı. Uygulama her açılışta kimlik doğrulaması isteyecek.'
      : 'Biyometrik kilit kapatıldı.');
    setBusy(false);
  };

  return (
    <Card title="🫆 Biyometrik Kilit">
      <Text style={styles.hint}>
        Açıkken uygulama her açılışta parmak izi / yüz tanıma ister. Biyometrik
        verilerin cihazından asla çıkmaz ve uygulama tarafından kaydedilmez;
        doğrulama başarısız olursa şifreyle giriş her zaman mümkündür.
      </Text>
      <View style={styles.switchRow}>
        <Text style={styles.switchLabel}>{enabled ? 'Açık' : 'Kapalı'}</Text>
        <Switch value={enabled} onValueChange={toggle} disabled={busy} trackColor={{ true: colors.primary }} />
      </View>
      {msg && <Text style={styles.ok}>{msg}</Text>}
    </Card>
  );
}

const EVENT_LABELS = {
  login_success: '✅ Giriş yapıldı',
  login_failed: '⚠️ Başarısız giriş denemesi',
  login_unverified: '⚠️ Doğrulanmamış e-posta ile giriş denemesi',
  logout: '👋 Çıkış yapıldı',
  logout_all: '🚪 Diğer oturumlar kapatıldı',
  session_revoked: '📵 Bir cihazın oturumu kapatıldı',
  password_changed: '🔑 Şifre değiştirildi',
  email_change_requested: '📧 E-posta değişikliği istendi',
  reauth_failed: '⚠️ Hatalı şifreyle doğrulama denemesi',
  refresh_denied: '⛔ Kapatılmış oturumla erişim denemesi',
  register: '🆕 Hesap oluşturuldu',
  account_delete_started: '🗑️ Hesap silme başlatıldı',
};

export default function SecuritySettingsScreen() {
  const { user } = useAuth();

  // Şifre değiştirme
  const [curPw, setCurPw] = useState('');
  const [newPw, setNewPw] = useState('');
  const [pwBusy, setPwBusy] = useState(false);
  const [pwOk, setPwOk] = useState(null);
  const [pwErr, setPwErr] = useState(null);
  const changePw = async () => {
    setPwOk(null); setPwErr(null); setPwBusy(true);
    try {
      const r = await api.changePassword({ currentPassword: curPw, newPassword: newPw });
      setPwOk(r.message || 'Şifren değiştirildi.');
      setCurPw(''); setNewPw('');
    } catch (e) { setPwErr(e.message); } finally { setPwBusy(false); }
  };

  // E-posta değiştirme
  const [newEmail, setNewEmail] = useState('');
  const [emailPw, setEmailPw] = useState('');
  const [emBusy, setEmBusy] = useState(false);
  const [emOk, setEmOk] = useState(null);
  const [emErr, setEmErr] = useState(null);
  const changeEmail = async () => {
    setEmOk(null); setEmErr(null); setEmBusy(true);
    try {
      const r = await api.changeEmail({ newEmail: newEmail.trim(), currentPassword: emailPw });
      setEmOk(r.message);
      setNewEmail(''); setEmailPw('');
      refreshUser();
    } catch (e) { setEmErr(e.message); } finally { setEmBusy(false); }
  };

  // Diğer oturumları kapatma
  const [allBusy, setAllBusy] = useState(false);
  const [allMsg, setAllMsg] = useState(null);
  const closeOthers = async () => {
    setAllBusy(true); setAllMsg(null);
    try {
      const r = await api.logoutAll();
      setAllMsg(`${r.closed} oturum kapatıldı. Bu cihaz açık kaldı.`);
    } catch (e) { setAllMsg(e.message); } finally { setAllBusy(false); }
  };

  // Güvenlik olayları
  const [events, setEvents] = useState(null);
  useEffect(() => {
    let alive = true;
    api.securityEvents().then((r) => { if (alive) setEvents(r.events || []); }).catch(() => { if (alive) setEvents([]); });
    return () => { alive = false; };
  }, []);

  return (
    <ScreenBackdrop>
      <ScrollView style={styles.container} contentContainerStyle={{ padding: spacing.lg }}>
        <Text style={styles.title}>Güvenlik Ayarları</Text>

        <BiometricCard />

        <Card title="🔑 Şifre Değiştir">
          <Text style={styles.hint}>Onaylandığında diğer cihazlardaki oturumların güvenlik için kapatılır.</Text>
          <PasswordField label="Mevcut şifre" value={curPw} onChangeText={setCurPw} placeholder="••••••••" />
          <PasswordField label="Yeni şifre" value={newPw} onChangeText={setNewPw} showStrength />
          <Feedback ok={pwOk} err={pwErr} />
          <TouchableOpacity style={[styles.btn, pwBusy && styles.btnBusy]} onPress={changePw} disabled={pwBusy}>
            {pwBusy ? <ActivityIndicator color={colors.bg} /> : <Text style={styles.btnTxt}>Şifreyi Değiştir</Text>}
          </TouchableOpacity>
        </Card>

        <Card title="📧 E-posta Değiştir">
          <Text style={styles.hint}>Mevcut adres: {user?.email || '—'}. Değişiklik, yeni adrese gelen doğrulama bağlantısıyla tamamlanır.</Text>
          <Text style={styles.label}>Yeni e-posta</Text>
          <TextInput style={styles.input} value={newEmail} onChangeText={setNewEmail} autoCapitalize="none" keyboardType="email-address" placeholder="yeni@ornek.com" placeholderTextColor={colors.textMuted} />
          <PasswordField label="Mevcut şifre" value={emailPw} onChangeText={setEmailPw} placeholder="••••••••" />
          <Feedback ok={emOk} err={emErr} />
          <TouchableOpacity style={[styles.btn, emBusy && styles.btnBusy]} onPress={changeEmail} disabled={emBusy}>
            {emBusy ? <ActivityIndicator color={colors.bg} /> : <Text style={styles.btnTxt}>Doğrulama Bağlantısı Gönder</Text>}
          </TouchableOpacity>
        </Card>

        <Card title="🚪 Diğer Oturumları Kapat">
          <Text style={styles.hint}>Şüpheli bir durum fark edersen bu cihaz dışındaki tüm oturumları tek dokunuşla kapatabilirsin.</Text>
          {allMsg && <Text style={styles.ok}>{allMsg}</Text>}
          <TouchableOpacity style={[styles.btn, styles.btnWarn, allBusy && styles.btnBusy]} onPress={closeOthers} disabled={allBusy}>
            {allBusy ? <ActivityIndicator color={colors.red} /> : <Text style={styles.btnWarnTxt}>Diğer Tüm Oturumları Kapat</Text>}
          </TouchableOpacity>
        </Card>

        <Card title="🕓 Son Güvenlik Olayları">
          {events === null && <ActivityIndicator color={colors.primary} />}
          {events?.length === 0 && <Text style={styles.hint}>Henüz kayıtlı olay yok.</Text>}
          {(events || []).map((e, i) => (
            <View key={i} style={styles.eventRow}>
              <Text style={styles.eventLabel}>{EVENT_LABELS[e.event] || e.event}</Text>
              <Text style={styles.eventMeta}>
                {new Date(e.created_at).toLocaleString('tr-TR')}{e.ip ? ` · ${e.ip}` : ''}
              </Text>
            </View>
          ))}
        </Card>
      </ScrollView>
    </ScreenBackdrop>
  );
}

const styles = StyleSheet.create({
  container: { flex: 1, backgroundColor: 'transparent' },
  title: { color: colors.text, fontSize: 22, fontWeight: '800', marginBottom: spacing.lg },
  card: { backgroundColor: colors.card, borderRadius: radius.md, borderWidth: 1, borderColor: colors.border, padding: spacing.md, marginBottom: spacing.md },
  cardTitle: { color: colors.text, fontSize: 15.5, fontWeight: '800', marginBottom: 8 },
  hint: { color: colors.textMuted, fontSize: 12.5, lineHeight: 18, marginBottom: spacing.md },
  label: { color: colors.textMuted, fontSize: 12, fontWeight: '700', marginBottom: 6 },
  input: { backgroundColor: colors.cardAlt, color: colors.text, borderRadius: radius.md, paddingHorizontal: 14, paddingVertical: 12, fontSize: 15, borderWidth: 1, borderColor: colors.border, marginBottom: spacing.md },
  btn: { backgroundColor: colors.primary, paddingVertical: 13, borderRadius: radius.md, alignItems: 'center', marginTop: 4 },
  btnBusy: { opacity: 0.6 },
  btnTxt: { color: colors.bg, fontSize: 14.5, fontWeight: '800' },
  btnWarn: { backgroundColor: 'transparent', borderWidth: 1, borderColor: colors.red },
  btnWarnTxt: { color: colors.red, fontSize: 14.5, fontWeight: '800' },
  ok: { color: colors.green, fontSize: 13, marginBottom: 8, fontWeight: '600' },
  err: { color: colors.red, fontSize: 13, marginBottom: 8, fontWeight: '600' },
  switchRow: { flexDirection: 'row', alignItems: 'center', justifyContent: 'space-between' },
  switchLabel: { color: colors.text, fontSize: 14, fontWeight: '800' },
  eventRow: { borderBottomWidth: 1, borderBottomColor: colors.border, paddingVertical: 8 },
  eventLabel: { color: colors.text, fontSize: 13, fontWeight: '700' },
  eventMeta: { color: colors.textMuted, fontSize: 11.5, marginTop: 2 },
});

File: lib/features/profile/security_settings_screen.dart
// KAYNAK: app/src/screens/SecuritySettingsScreen.js — BİREBİR çeviri.
//
// GÜVENLİK AYARLARI — şifre değiştir · e-posta değiştir · diğer oturumları
// kapat · son güvenlik olayları.
//
//   • Şifre ve e-posta değişikliği MEVCUT ŞİFRE ile yeniden doğrulama ister
//     (sunucu da ayrıca zorunlu kılar — arayüz atlatılamaz).
//   • Şifre değişince diğer cihazlardaki oturumlar sunucuda kapatılır.
//   • Olay listesinde şifre/token benzeri hassas veri YOKTUR (sunucu süzer).

import 'dart:async';

import 'package:flutter/material.dart';

import '../../core/auth.dart' as auth;
import '../../core/network/api_client.dart';
import '../../core/security/biometric_lock.dart';
import '../../core/theme/tokens.dart';
import '../../widgets/screen_backdrop.dart';
import 'auth_screens.dart' show PasswordField;

const Map<String, String> _eventLabels = {
  'login_success': '✅ Giriş yapıldı',
  'login_failed': '⚠️ Başarısız giriş denemesi',
  'login_unverified': '⚠️ Doğrulanmamış e-posta ile giriş denemesi',
  'logout': '👋 Çıkış yapıldı',
  'logout_all': '🚪 Diğer oturumlar kapatıldı',
  'session_revoked': '📵 Bir cihazın oturumu kapatıldı',
  'password_changed': '🔑 Şifre değiştirildi',
  'email_change_requested': '📧 E-posta değişikliği istendi',
  'reauth_failed': '⚠️ Hatalı şifreyle doğrulama denemesi',
  'refresh_denied': '⛔ Kapatılmış oturumla erişim denemesi',
  'register': '🆕 Hesap oluşturuldu',
  'account_delete_started': '🗑️ Hesap silme başlatıldı',
};

class SecuritySettingsScreen extends StatefulWidget {
  const SecuritySettingsScreen({super.key});

  @override
  State<SecuritySettingsScreen> createState() => _SecuritySettingsScreenState();
}

class _SecuritySettingsScreenState extends State<SecuritySettingsScreen> {
  // Şifre değiştirme
  final _curPw = TextEditingController();
  final _newPw = TextEditingController();
  bool _pwBusy = false;
  String? _pwOk;
  String? _pwErr;

  // E-posta değiştirme
  final _newEmail = TextEditingController();
  final _emailPw = TextEditingController();
  bool _emBusy = false;
  String? _emOk;
  String? _emErr;

  // Diğer oturumları kapatma
  bool _allBusy = false;
  String? _allMsg;

  // Güvenlik olayları
  List<dynamic>? _events;

  @override
  void initState() {
    super.initState();
    api
        .securityEvents()
        .then((r) {
          if (!mounted) return;
          setState(
            () => _events = (r is Map ? r['events'] as List? : null) ?? [],
          );
        })
        .catchError((_) {
          if (mounted) setState(() => _events = []);
        });
  }

  @override
  void dispose() {
    _curPw.dispose();
    _newPw.dispose();
    _newEmail.dispose();
    _emailPw.dispose();
    super.dispose();
  }

  Future<void> _changePw() async {
    setState(() {
      _pwOk = null;
      _pwErr = null;
      _pwBusy = true;
    });
    try {
      final r = await api.changePassword({
        'currentPassword': _curPw.text,
        'newPassword': _newPw.text,
      });
      if (!mounted) return;
      setState(() {
        _pwOk =
            (r is Map ? r['message'] as String? : null) ??
            'Şifren değiştirildi.';
        _curPw.clear();
        _newPw.clear();
      });
    } catch (e) {
      if (mounted) setState(() => _pwErr = '$e');
    } finally {
      if (mounted) setState(() => _pwBusy = false);
    }
  }

  Future<void> _changeEmail() async {
    setState(() {
      _emOk = null;
      _emErr = null;
      _emBusy = true;
    });
    try {
      final r = await api.changeEmail({
        'newEmail': _newEmail.text.trim(),
        'currentPassword': _emailPw.text,
      });
      if (!mounted) return;
      setState(() {
        _emOk = r is Map ? r['message'] as String? : null;
        _newEmail.clear();
        _emailPw.clear();
      });
      // Kaynakta beklenmeden çağrılıyor — e-posta doğrulanana kadar profil
      // zaten değişmez, bu yüzden sonucu beklemek gerekmiyor.
      unawaited(auth.refreshUser());
    } catch (e) {
      if (mounted) setState(() => _emErr = '$e');
    } finally {
      if (mounted) setState(() => _emBusy = false);
    }
  }

  Future<void> _closeOthers() async {
    setState(() {
      _allBusy = true;
      _allMsg = null;
    });
    try {
      final r = await api.logoutAll();
      final kapanan = r is Map ? r['closed'] : null;
      if (mounted) {
        setState(
          () => _allMsg = '$kapanan oturum kapatıldı. Bu cihaz açık kaldı.',
        );
      }
    } catch (e) {
      if (mounted) setState(() => _allMsg = '$e');
    } finally {
      if (mounted) setState(() => _allBusy = false);
    }
  }

  /// Kaynakta `new Date(created_at).toLocaleString('tr-TR')` → "9.08.2026 10:34:00"
  String _olayZamani(Object? iso) {
    final d = DateTime.tryParse('$iso')?.toLocal();
    if (d == null) return '—';
    String p(int n) => n.toString().padLeft(2, '0');
    return '${d.day}.${p(d.month)}.${d.year} ${p(d.hour)}:${p(d.minute)}:${p(d.second)}';
  }

  @override
  Widget build(BuildContext context) => ValueListenableBuilder<auth.AuthState>(
    valueListenable: auth.authState,
    builder: (context, s, _) => Scaffold(
      backgroundColor: AppColors.bg,
      appBar: AppBar(title: const Text('Güvenlik Ayarları')),
      body: ScreenBackdrop(
        child: ListView(
          padding: const EdgeInsets.all(Spacing.lg),
          children: [
            const Text(
              'Güvenlik Ayarları',
              style: TextStyle(
                color: AppColors.text,
                fontSize: 22,
                fontWeight: AppFont.heavy,
              ),
            ),
            const SizedBox(height: Spacing.lg),

            const _BiometricCard(),

            _Kart(
              baslik: '🔑 Şifre Değiştir',
              children: [
                const _Ipucu(
                  'Onaylandığında diğer cihazlardaki oturumların güvenlik için kapatılır.',
                ),
                PasswordField(
                  label: 'Mevcut şifre',
                  controller: _curPw,
                  placeholder: '••••••••',
                ),
                PasswordField(
                  label: 'Yeni şifre',
                  controller: _newPw,
                  showStrength: true,
                ),
                _Geribildirim(ok: _pwOk, err: _pwErr),
                _Dugme(
                  metin: 'Şifreyi Değiştir',
                  busy: _pwBusy,
                  onTap: _changePw,
                ),
              ],
            ),

            _Kart(
              baslik: '📧 E-posta Değiştir',
              children: [
                _Ipucu(
                  'Mevcut adres: ${s.user?['email'] ?? '—'}. Değişiklik, yeni adrese gelen doğrulama bağlantısıyla tamamlanır.',
                ),
                const Text(
                  'Yeni e-posta',
                  style: TextStyle(
                    color: AppColors.textMuted,
                    fontSize: 12,
                    fontWeight: AppFont.bold,
                  ),
                ),
                const SizedBox(height: 6),
                TextField(
                  controller: _newEmail,
                  keyboardType: TextInputType.emailAddress,
                  autocorrect: false,
                  style: const TextStyle(color: AppColors.text, fontSize: 15),
                  decoration: InputDecoration(
                    hintText: 'yeni@ornek.com',
                    hintStyle: const TextStyle(color: AppColors.textMuted),
                    filled: true,
                    fillColor: AppColors.cardAlt,
                    isDense: true,
                    contentPadding: const EdgeInsets.symmetric(
                      horizontal: 14,
                      vertical: 12,
                    ),
                    border: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(AppRadius.md),
                      borderSide: const BorderSide(color: AppColors.border),
                    ),
                    enabledBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(AppRadius.md),
                      borderSide: const BorderSide(color: AppColors.border),
                    ),
                    focusedBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(AppRadius.md),
                      borderSide: const BorderSide(color: AppColors.border),
                    ),
                  ),
                ),
                const SizedBox(height: Spacing.md),
                PasswordField(
                  label: 'Mevcut şifre',
                  controller: _emailPw,
                  placeholder: '••••••••',
                ),
                _Geribildirim(ok: _emOk, err: _emErr),
                _Dugme(
                  metin: 'Doğrulama Bağlantısı Gönder',
                  busy: _emBusy,
                  onTap: _changeEmail,
                ),
              ],
            ),

            _Kart(
              baslik: '🚪 Diğer Oturumları Kapat',
              children: [
                const _Ipucu(
                  'Şüpheli bir durum fark edersen bu cihaz dışındaki tüm oturumları tek dokunuşla kapatabilirsin.',
                ),
                if (_allMsg != null) _Onay(_allMsg!),
                _Dugme(
                  metin: 'Diğer Tüm Oturumları Kapat',
                  busy: _allBusy,
                  onTap: _closeOthers,
                  uyari: true,
                ),
              ],
            ),

            _Kart(
              baslik: '🕓 Son Güvenlik Olayları',
              children: [
                if (_events == null)
                  const Center(
                    child: SizedBox(
                      width: 20,
                      height: 20,
                      child: CircularProgressIndicator(
                        strokeWidth: 2,
                        color: AppColors.primary,
                      ),
                    ),
                  ),
                if (_events != null && _events!.isEmpty)
                  const _Ipucu('Henüz kayıtlı olay yok.'),
                for (final e in _events ?? const [])
                  _olaySatiri((e as Map).cast<String, dynamic>()),
              ],
            ),
          ],
        ),
      ),
    ),
  );

  Widget _olaySatiri(Map<String, dynamic> e) {
    final ip = e['ip'];
    final ipVar = ip != null && '$ip'.isNotEmpty;
    return Container(
      padding: const EdgeInsets.symmetric(vertical: 8),
      decoration: const BoxDecoration(
        border: Border(bottom: BorderSide(color: AppColors.border)),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
            // Bilinmeyen olay TÜRÜ ham anahtarıyla gösterilir — sunucu yeni
            // bir olay eklediğinde satır KAYBOLMAZ (kaynak aynen).
            _eventLabels['${e['event']}'] ?? '${e['event']}',
            style: const TextStyle(
              color: AppColors.text,
              fontSize: 13,
              fontWeight: AppFont.bold,
            ),
          ),
          Padding(
            padding: const EdgeInsets.only(top: 2),
            child: Text(
              '${_olayZamani(e['created_at'])}${ipVar ? ' · $ip' : ''}',
              style: const TextStyle(
                color: AppColors.textMuted,
                fontSize: 11.5,
              ),
            ),
          ),
        ],
      ),
    );
  }
}

/// Biyometrik kilit — yalnız destekleyen cihazlarda görünür (web'de hiç).
/// Aç/kapat için önce cihazın biyometrik doğrulaması istenir; böylece kilidi
/// yalnız cihaz sahibi değiştirebilir. Biyometrik veri uygulamaya girmez.
class _BiometricCard extends StatefulWidget {
  const _BiometricCard();

  @override
  State<_BiometricCard> createState() => _BiometricCardState();
}

class _BiometricCardState extends State<_BiometricCard> {
  bool _supported = false;
  bool _enabled = false;
  bool _busy = false;
  String? _msg;

  @override
  void initState() {
    super.initState();
    () async {
      final s = await biometricsSupported();
      if (!mounted) return;
      setState(() => _supported = s);
      if (s) {
        final e = await isBioLockEnabled();
        if (mounted) setState(() => _enabled = e);
      }
    }();
  }

  Future<void> _toggle(bool next) async {
    setState(() {
      _busy = true;
      _msg = null;
    });
    // değişiklik için kimlik doğrulaması şart
    final outcome = await authenticate();
    if (outcome != 'unlocked') {
      if (mounted) {
        setState(() {
          _msg = 'Doğrulama başarısız — ayar değiştirilmedi.';
          _busy = false;
        });
      }
      return;
    }
    await setBioLockEnabled(next);
    if (!mounted) return;
    setState(() {
      _enabled = next;
      _msg = next
          ? 'Biyometrik kilit açıldı. Uygulama her açılışta kimlik doğrulaması isteyecek.'
          : 'Biyometrik kilit kapatıldı.';
      _busy = false;
    });
  }

  @override
  Widget build(BuildContext context) {
    if (!_supported) return const SizedBox.shrink();
    return _Kart(
      baslik: '🫆 Biyometrik Kilit',
      children: [
        const _Ipucu(
          'Açıkken uygulama her açılışta parmak izi / yüz tanıma ister. Biyometrik '
          'verilerin cihazından asla çıkmaz ve uygulama tarafından kaydedilmez; '
          'doğrulama başarısız olursa şifreyle giriş her zaman mümkündür.',
        ),
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text(
              _enabled ? 'Açık' : 'Kapalı',
              style: const TextStyle(
                color: AppColors.text,
                fontSize: 14,
                fontWeight: AppFont.heavy,
              ),
            ),
            Switch(
              value: _enabled,
              onChanged: _busy ? null : _toggle,
              activeTrackColor: AppColors.primary,
            ),
          ],
        ),
        if (_msg != null) _Onay(_msg!),
      ],
    );
  }
}

class _Kart extends StatelessWidget {
  const _Kart({required this.baslik, required this.children});

  final String baslik;
  final List<Widget> children;

  @override
  Widget build(BuildContext context) => Container(
    margin: const EdgeInsets.only(bottom: Spacing.md),
    padding: const EdgeInsets.all(Spacing.md),
    decoration: BoxDecoration(
      color: AppColors.card,
      borderRadius: BorderRadius.circular(AppRadius.md),
      border: Border.all(color: AppColors.border),
    ),
    child: Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          baslik,
          style: const TextStyle(
            color: AppColors.text,
            fontSize: 15.5,
            fontWeight: AppFont.heavy,
          ),
        ),
        const SizedBox(height: 8),
        ...children,
      ],
    ),
  );
}

class _Ipucu extends StatelessWidget {
  const _Ipucu(this.metin);

  final String metin;

  @override
  Widget build(BuildContext context) => Padding(
    padding: const EdgeInsets.only(bottom: Spacing.md),
    child: Text(
      metin,
      style: const TextStyle(
        color: AppColors.textMuted,
        fontSize: 12.5,
        height: 18 / 12.5,
      ),
    ),
  );
}

class _Onay extends StatelessWidget {
  const _Onay(this.metin);

  final String metin;

  @override
  Widget build(BuildContext context) => Padding(
    padding: const EdgeInsets.only(bottom: 8),
    child: Text(
      metin,
      style: const TextStyle(
        color: AppColors.green,
        fontSize: 13,
        fontWeight: AppFont.semibold,
      ),
    ),
  );
}

class _Geribildirim extends StatelessWidget {
  const _Geribildirim({this.ok, this.err});

  final String? ok;
  final String? err;

  @override
  Widget build(BuildContext context) {
    if (ok != null) return _Onay(ok!);
    if (err != null) {
      return Padding(
        padding: const EdgeInsets.only(bottom: 8),
        child: Text(
          err!,
          style: const TextStyle(
            color: AppColors.red,
            fontSize: 13,
            fontWeight: AppFont.semibold,
          ),
        ),
      );
    }
    return const SizedBox.shrink();
  }
}

class _Dugme extends StatelessWidget {
  const _Dugme({
    required this.metin,
    required this.busy,
    required this.onTap,
    this.uyari = false,
  });

  final String metin;
  final bool busy;
  final VoidCallback onTap;
  final bool uyari;

  @override
  Widget build(BuildContext context) => Opacity(
    opacity: busy ? 0.6 : 1,
    child: GestureDetector(
      onTap: busy ? null : onTap,
      child: Container(
        margin: const EdgeInsets.only(top: 4),
        padding: const EdgeInsets.symmetric(vertical: 13),
        alignment: Alignment.center,
        decoration: BoxDecoration(
          color: uyari ? null : AppColors.primary,
          borderRadius: BorderRadius.circular(AppRadius.md),
          border: uyari ? Border.all(color: AppColors.red) : null,
        ),
        child: busy
            ? SizedBox(
                width: 18,
                height: 18,
                child: CircularProgressIndicator(
                  strokeWidth: 2,
                  color: uyari ? AppColors.red : AppColors.bg,
                ),
              )
            : Text(
                metin,
                style: TextStyle(
                  color: uyari ? AppColors.red : AppColors.bg,
                  fontSize: 14.5,
                  fontWeight: AppFont.heavy,
                ),
              ),
      ),
    ),
  );
}
